packages/block-serialization-default-parser/parser.php

```diff
@@ -5,551 +5,7 @@
-		return new WP_Block_Parser_Block( null, $this->empty_attrs, array(), $innerHTML, array( $innerHTML ) );
-	}
-
-	/**
-	 * Pushes a length of text from the input document
-	 * to the output list as a freeform block.
-	 *
-	 * @internal
-	 * @since 5.0.0
-	 * @param null $length how many bytes of document text to output.
-	 */
-	public function add_freeform( $length = null ) {
-		$length = $length ? $length : strlen( $this->document ) - $this->offset;
-
-		if ( 0 === $length ) {
-			return;
-		}
-
-		$this->output[] = (array) $this->freeform( substr( $this->document, $this->offset, $length ) );
-	}
-
-	/**
-	 * Given a block structure from memory pushes
-	 * a new block to the output list.
-	 *
-	 * @internal
-	 * @since 5.0.0
-	 * @param WP_Block_Parser_Block $block        The block to add to the output.
-	 * @param int                   $token_start  Byte offset into the document where the first token for the block starts.
-	 * @param int                   $token_length Byte length of entire block from start of opening token to end of closing token.
-	 * @param int|null              $last_offset  Last byte offset into document if continuing form earlier output.
-	 */
-	public function add_inner_block( WP_Block_Parser_Block $block, $token_start, $token_length, $last_offset = null ) {
-		$parent                       = $this->stack[ count( $this->stack ) - 1 ];
-		$parent->block->innerBlocks[] = (array) $block;
-		$html                         = substr( $this->document, $parent->prev_offset, $token_start - $parent->prev_offset );
-
-		if ( ! empty( $html ) ) {
-			$parent->block->innerHTML     .= $html;
-			$parent->block->innerContent[] = $html;
-		}
-
-		$parent->block->innerContent[] = null;
-		$parent->prev_offset           = $last_offset ? $last_offset : $token_start + $token_length;
-	}
-
-	/**
-	 * Pushes the top block from the parsing stack to the output list.
-	 *
-	 * @internal
-	 * @since 5.0.0
-	 * @param int|null $end_offset byte offset into document for where we should stop sending text output as HTML.
-	 */
-	public function add_block_from_stack( $end_offset = null ) {
-		$stack_top   = array_pop( $this->stack );
-		$prev_offset = $stack_top->prev_offset;
-
-		$html = isset( $end_offset )
-			? substr( $this->document, $prev_offset, $end_offset - $prev_offset )
-			: substr( $this->document, $prev_offset );
-
-		if ( ! empty( $html ) ) {
-			$stack_top->block->innerHTML     .= $html;
-			$stack_top->block->innerContent[] = $html;
-		}
-
-		if ( isset( $stack_top->leading_html_start ) ) {
-			$this->output[] = (array) $this->freeform(
-				substr(
-					$this->document,
-					$stack_top->leading_html_start,
-					$stack_top->token_start - $stack_top->leading_html_start
-				)
-			);
-		}
-
-		$this->output[] = (array) $stack_top->block;
-	}
-}
+// Require files.
+require_once __DIR__ . '/class-wp-block-parser-block.php';
+require_once __DIR__ . '/class-wp-block-parser-frame.php';
+require_once __DIR__ . '/class-wp-block-parser.php';
```

Comment:
although I still am of the opinion that this PR is worth closing without merging, I wanted to point out that @hellofromtonya rejected HTML API work for including files like this, instead of doing everything manually in `load.php`/`wp-settings.php`

but again, I still believe it's best to exclude this file from the one-class-per-file rule because of its close correspondence to the other parser implementations and because of the the role the helper classes serve.

Reply:
@dmsnell I am not why this should be a special case. Having the one in the same file is confusing and makes it harder to find the class. It also making autoloading core classes worse as well.

Reply:
what's the problem with autoloading core classes?

> I am not why this should be a special case

I've shared a few times:
 - these helper classes are specifically empty classes serving as records. they contain no behavior
 - they directly mirror the code in the JS and Java versions of the parser
 - they have no valuable use outside of the parser and are not meant to be loaded in any other context

Reply:
@spacedmonkey I can understand if you don't see this as a benefit, but there's a benefit there. maybe you don't value it as much, but it's not like there's no benefit.

> Autoloading is a project to load class files only when they are used. This saves on PHP memory, as classes are not loaded into memory that are not used. With all three classes in this one file, that is impossible.

This just hearkens back to the reasoning I gave: there's zero purpose to ever load any of the helper classes outside the parser, and there's zero cases where the parser loads without loading the helper classes, so there's no situation in which this desired behavior would even come into play with the parser.

Are you saying that having two or more classes in a single file prevents any autoloading from working within WordPress? or that having these classes in the same file _here_ prevents autoloading _these_ classes?

if it's the latter then the principle has zero application to this case, other than in a purely theoretical sense. there's no memory savings by splitting these as they will _always_ load together, and there will never be a time when they should load separately.

Reply:
@dmsnell Unit tests is a good example of where you might want to load these files apart. 

In my testing `WP_Block_Parser_Frame` did not load on every page request. 

I think we are going to have to agree to disagree on this one.

Reply:
>but again, I still believe it's best to exclude this file from the one-class-per-file rule because of its close correspondence to the other parser implementations and because of the the role the helper classes serve.

The WordPress Coding Standard is one class/interface/trait per file. This is documented in Core's Handbook. https://developer.wordpress.org/coding-standards/wordpress-coding-standards/php/#only-one-object-structure-class-interface-trait-per-file

@dmsnell though I see your reasoning with close correspondence, I disagree with excluding the file from the rule. All new code should meet Core's coding standards.

Reply:
```php
// Require files.
require_once __DIR__ . '/class-wp-block-parser-block.php';
require_once __DIR__ . '/class-wp-block-parser-frame.php';
require_once __DIR__ . '/class-wp-block-parser.php';
```

>I wanted to point out that @hellofromtonya rejected HTML API work for including files like this, instead of doing everything manually in `load.php/wp-settings.php`

The loading of these files should be where all other files are loaded, and not in the `parser.php` file itself. Why? Consistency with how Core loads its file.

Reply:
> The loading of these files should be where all other files are loaded, and not in the `parser.php` file itself. Why? Consistency with how Core loads its file.

I agree 100%. This was done here like that for backwards-compatibility.
If a plugin or 3rd-party includes the `parser.php` file, it should continue to work - which is why instead of removing the file since it's no longer needed, I include the classes there 👍